### ISSUE TITLE
Utility of updated map.proto?

### ISSUE THREAD

**Issue by MoritzNekolla** [waymo-research/waymo-open-dataset#389]
Hey all,

with the updated map.proto file, we have access to `lane.entry/exit` lanes, `lane.neighbors` and `lane.borders`. Given the [goal ](https://github.com/waymo-research/waymo-open-dataset/issues/349) to retrieve a lane graph (where a lane consists of a left and right lane), this sounded promising. However, it remains difficult. For instance, sticking together consecutive lanes works perfectly but assigning the left/right neighbors is rough, since lanes tend to have multiple neighbors. The other possibility is to first create lane segments and later stick them together. Both methods do not work properly. Is there a trick on how to cope with your newly provided data?

Greetings

**Comment by scott-ettinger** [comment]
Could you describe further what you are trying to do?  I'm not sure I fully understand the issue.

**Comment by MoritzNekolla** [comment]
Hey Scott,

thanks for replying that fast.

I want to develop a converter from waymo to [liblanelet](https://github.com/fzi-forschungszentrum-informatik/liblanelet). In liblanelet every lane consist of a left and right polyline. With the updated map.proto I thought retrieving such structure should be possible. However, I am struggling to make it work.
My first attempt was to stick consecutive lanes together with `lane.entry_lanes` and `lane.exit_lanes` which worked. The issue is with `lane.left_neighbors` and `lane.right_neighbors`. Both contain road_edge and road_line which do not have entry/exit_lanes. Therefore, I can't glue together the left/right polylines since the original lane has multiple neighbors.
In a nutshell, it feels a bit odd working with the newly provided data but this could be because of my wrong understanding of the data. Thus, my question is if you could clarify the idea behind the data structure or how to solve my problem.

**Comment by scott-ettinger** [comment]
Hi,
The left and right neighbors should only contain other lane centers (which do have entry and exit lanes).   Are you seeing otherwise?  Only the lane boundaries should contain road edges and road lines.  I'm not sure how liblanelet represents its graph, but at each point in the lane you should be able to add a connection to its lane neighbors.  Is the problem that it requires only a single connection for left and right?  If that is the case you could select only the closest of the neighbor points.  The data should lend itself well to a graph structure, we use this particular format to save space. If you have a specific example I can take a look.

**Comment by MoritzNekolla** [comment]
Here is an example:
Given the center lane 180 I want to extract left/right borders. Therefore I run this provisory code chunk:
![grafik](https://user-images.githubusercontent.com/65892179/142726068-ba3b15cd-8f1f-4f04-bd77-7bdbc076be6a.png)
The result is kind of weird because these borders appear not as one would expect. (Especially the purple one)

**Comment by scott-ettinger** [comment]
Do you have a scenario ID for this example?

**Comment by MoritzNekolla** [comment]
```
File name: uncompressed_scenario_validation_validation.tfrecord-00018-of-00150
scenario_id: 978e95dd1af00247
```
Or the "first" scenario when loading the file

**Comment by scott-ettinger** [comment]
Apologies for not getting back to you on this.  I will look into this shortly.

**Comment by MoritzNekolla** [comment]
Hey @scott-ettinger,

did you take a look? No pressure, but maybe you forgot about it :)

**Comment by scott-ettinger** [comment]
Apologies for the very late response.  I thought I had responded to this. I now have legal approval to share some visualizations with you.  Here are visualizations of the left and right boundaries of the lane feature with ID 180 from the scenario you listed.  Please note that the visualizations are provided here under the Waymo Dataset License Agreement for Non-Commercial Use.

For more context, I have also included a visualization of the full left and right neighbors for the same area below as well.  Unfortunately I cannot share the code for this visualization at this time.  I hope to be able to add visualization code in colab for the next release. I cannot tell what might be causing the issue from your code above. I am happy to answer any questions to try to track it down. 

Left Boundaries for lane feature with ID 180:
![image](https://user-images.githubusercontent.com/82909755/151225141-5ffe7264-46da-4dc7-aa89-cd40d6d2fa4e.png)

Right Boundaries for lane feature with ID 180:
![image](https://user-images.githubusercontent.com/82909755/151224915-f1307ccf-ff60-4cc4-b1d2-44ab9ecbb7ca.png)

Left Boundaries for all Lanes: 
![image](https://user-images.githubusercontent.com/82909755/151225858-35ea84d5-d9c7-4e20-bd2f-fb6799fc7c49.png)

Right Boundaries for all Lanes:
![image](https://user-images.githubusercontent.com/82909755/151225901-0b103a5b-0c04-4d2e-a7f8-107996c8fe67.png)
